File: src/Kattis/COMP321/Assignments/A3/C.cpp
#include <iostream>

using namespace std;

int main() {
    int numCalls, numIntervals;
    while(true) {
        cin >> numCalls;
        cin >> numIntervals;
        if (numCalls == 0 && numIntervals == 0) {
            break;
        }

        int calls[numCalls][2];
        for (int i = 0; i < numCalls; i++) {
            int start, duration;
            cin >> start; // Ignore source
            cin >> start; // Ignore dest
            cin >> start;
            cin >> duration;
            calls[i][0] = start;
            calls[i][1] = start + duration;
        }

        int operators, cStart, cEnd;
        for (int i = 0; i < numIntervals; i++) {
            int iStart, iDuration;
            cin >> iStart;
            cin >> iDuration;

            operators = 0;

            for (int j = 0; j < numCalls; j++) {
                cStart = calls[j][0];
                cEnd = calls[j][1];
                if (cStart <= iStart && iStart < cEnd) {
                    operators++;
                } else if (iStart <= cStart && cStart < iStart + iDuration) {
                    operators++;
                }
            }
            cout << operators << endl;
        }
    }
    return 0;
}

File: src/Kattis/COMP321/Assignments/A3/C.java
package Kattis.COMP321.A3;

import java.util.Scanner;

public class C {
    public static void main(String[] args) {
        telephones();
    }

    private static void telephones() {
        // Create scanner object
        Scanner scanner = new Scanner(System.in);

        // Variables
        int numCalls, numIntervals, operators, start, duration, startOfCall, endOfCall;
        String[] input;
        int[][] calls;

        // Get test case
        input = scanner.nextLine().split("\\s");

        // Get number of phone calls and intervals
        numCalls = Integer.parseInt(input[0]);
        numIntervals = Integer.parseInt(input[1]);

        while (numCalls != 0 && numIntervals != 0) {
            // Get phone calls
            calls = new int[numCalls][];
            for (int i = 0; i < numCalls; i++) {
                input = scanner.nextLine().split("\\s");
                start = Integer.parseInt(input[2]);
                duration = Integer.parseInt(input[3]);

                // Increment number of active phone calls during [start, start + duration[
                calls[i] = new int []{start, start + duration};
            }

            // Get interval values
            for (int i = 0; i < numIntervals; i++) {
                input = scanner.nextLine().split("\\s");
                start = Integer.parseInt(input[0]);
                duration = Integer.parseInt(input[1]);
                operators = 0;

                for (int[] call: calls) {
                    startOfCall = call[0];
                    endOfCall = call[1];
                    if (startOfCall < start && start < endOfCall)
                        operators++;
                    else if (start <= startOfCall && startOfCall < start + duration)
                        operators++;
                }

                // Output result
                System.out.println(operators);
            }

            // Get next test case
            input = scanner.nextLine().split("\\s");

            // Get number of phone calls and intervals
            numCalls = Integer.parseInt(input[0]);
            numIntervals = Integer.parseInt(input[1]);
        }
    }
}
